### ISSUE TITLE
Unable to Inject into 64-bit processes.

### ISSUE THREAD

**Issue by rlc5m8** [EasyHook/EasyHook#70]
I've successfully hooked/intercepted the output from various programs. Unfortunately, I cannot hook a 64-bit program.

Operating System: Windows 10 x64 Version 1511 Build 10586.104
Language of Main Program: C#
Language of DLL: C#

I've been compiling with .NET 4.5.2 Any CPU. I have tried:

.NET 4.0 Client Profile Any CPU
.NET 4.0 Any Cpu
.NET 4.0 x64
.NET 3.5 Client Profile (Edit: Used the correct binaries, and this still does not work)

EasyHook Library used: 2.7.5870

I've even tried using a minimal implementation of the DLL which fails.

The Error Received is:

There was an error while connecting to target:

System.ApplicationException: STATUS_INTERNAL_ERROR: Unknown error in injected C++ completion routine. (Code: 15)

   at EasyHook.RemoteHooking.CreateAndInject(String InEXEPath, String InCommandLine, Int32 InProcessCreationFlags, InjectionOptions InOptions, String InLibraryPath_x86, String InLibraryPath_x64, Int32& OutProcessId, Object[] InPassThruArgs)

   at Profile_Manager.MainWindow.OnLoad(Object sender, RoutedEventArgs e) in C:\Visual Studio 2015\Projects\Profile_Manager\Profile_Manager\MainWindow.xaml.cs:line 37

**Comment by rlc5m8** [comment]
This is my Minimal DLL which still fails on x64 load:

```
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Linq;
using System.Text;
using System.Threading;
using System.Threading.Tasks;
using System.Windows;
using EasyHook;

namespace HookTest
{
    public class FileMonInterface : MarshalByRefObject
    {
        public void IsInstalled(Int32 inClientPid) { Trace.WriteLine($"Dll has been installed in target {inClientPid}.\r\n"); Trace.Flush(); }
        public void OnCreateFile(Int32 inClientPid, String[] inFileNames)
        {
            foreach (String fileName in inFileNames)
            {
                Trace.WriteLine(fileName);
            }
        }

        public void ReportException(Exception inInfo) { Trace.WriteLine("The target process has reported an error:\r\n" + inInfo); }
        public void Ping() { Trace.WriteLine("Interface Pinged."); }
    }
    public class Main : IEntryPoint
    {

        public Main(RemoteHooking.IContext inContext, String inChannelName)
        {
        }
        public void Run(RemoteHooking.IContext inContext, String inChannelName)
        {
            MessageBox.Show("Hooked!");
            RemoteHooking.WakeUpProcess();
        }
    }
}
```

It works for every x32 executable I throw at it however

**Comment by rlc5m8** [comment]
I've tested with FileMon as well, it exhibits the same problem.

C:\Users\Basic\Downloads\EasyHook>FileMon "C:\Program Files (x86)\Internet Explorer\iexplore.exe"
Created and injected process 5092
<Press any key to exit>
FileMon has been installed in target 5092.


C:\Users\Basic\Downloads\EasyHook>FileMon "C:\Program Files\Internet Explorer\iexplore.exe"
There was an error while connecting to target:
System.ApplicationException: STATUS_INTERNAL_ERROR: Unknown error in injected C++ completion routine. (Code: 15)
   at EasyHook.RemoteHooking.CreateAndInject(String InEXEPath, String InCommandLine, Int32 InProcessCreationFlags, InjectionOptions InOptions, String InLibraryPath_x86, String InLibraryPath_x64, Int32& OutProcessId, Object[] InPassThruArgs)
   at FileMon.Program.Main(String[] args)

**Comment by spazzarama** [comment]
I'll see if I can reproduce this problem tmrw. Bit strange, I was definitely successfully hooking 64-bit processes recently.

**Comment by rlc5m8** [comment]
> I'll see if I can reproduce this problem tmrw. Bit strange, I was definitely successfully hooking 64-bit processes recently.

The man of the hour! Are you doing so on x64 Windows 10?

**Comment by spazzarama** [comment]
Yes, working with Win 10 here (last time I checked when I released 5870).

**Comment by rlc5m8** [comment]
Strange, the only thing that seems to matter is x86 vs x64

C:\Users\Basic\Downloads\EasyHook>FileMon C:/Windows/Notepad.exe
There was an error while connecting to target:
System.ApplicationException: STATUS_INTERNAL_ERROR: Unknown error in injected C++ completion routine. (Code: 15)
   at EasyHook.RemoteHooking.CreateAndInject(String InEXEPath, String InCommandLine, Int32 InProcessCreationFlags, InjectionOptions InOptions, String InLibraryPath_x86, String InLibraryPath_x64, Int32& OutProcessId, Object[] InPassThruArgs)
   at FileMon.Program.Main(String[] args)
<Press any key to exit>

C:\Users\Basic\Downloads\EasyHook>FileMon C:/Windows/SYSWOW64/Notepad.exe
Created and injected process 14996
<Press any key to exit>
FileMon has been installed in target 14996.

**Comment by spazzarama** [comment]
When I ran FileMon it worked as shown below (this is injecting into 64-bit notepad on Win10).

    Usage: FileMon %PID%
       or: FileMon PathToExecutable
    
    Please enter a process Id or path to executable: 11840
    Injected to process 11840
    <Press any key to exit>
    FileMon has been installed in target 11840.
    
    [11840:12100]: "\\.\pipe\DropboxPipe_1"
    [11840:12100]: "\\.\pipe\DropboxPipe_1"
    ...

I have managed to reproduce a similar sort of issue but this was only if I tried to inject a non-strongly named FileMonInject.dll and then tried to do it again with the correct version to the same process. I'll at least track this down in case it is related somehow - but using the binaries from the 5870 release it is working fine.

**Comment by spazzarama** [comment]
@rlc5m8 can you please run the fuslogvw.exe and log any .NET assembly load errors as the following error can relate to a number of issues within the EasyLoad32/64.dll related to loading EasyHook or the target injection assembly.

    There was an error while connecting to target:
    System.ApplicationException: STATUS_INTERNAL_ERROR: Unknown error in injected C++ 
    completion routine. (Code: 15)

**Comment by spazzarama** [comment]
@rlc5m8 can you also confirm that you are targetting "AnyCPU" and that you have unticked "Prefer 32-bit".

**Comment by spazzarama** [comment]
@rlc5m8 if you are using the solution from the repository, perhaps try building at the solution level - perhaps it is an assembly version issue?

**Comment by rlc5m8** [comment]
Sorry for the lack of replies, I've been sick all day. I'll start the additional testing now.

**Comment by rlc5m8** [comment]
Here's the output from fuslogvw when hooking 64bit Notepad.exe with FIleMon
"<meta http-equiv="Content-Type" content="charset=unicode-1-1-utf-8"><html><pre>
*** Assembly Binder Log Entry  (2/13/2016 @ 11:00:51 PM) ***

The operation was successful.
Bind result: hr = 0x0. The operation completed successfully.

Assembly manager loaded from:  C:\Windows\Microsoft.NET\Framework64\v4.0.30319\clr.dll
Running under executable  C:\Users\Basic\Downloads\EasyHook\FileMon.exe
--- A detailed error log follows. 

=== Pre-bind state information ===
LOG: DisplayName = System.Core, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089
 (Fully-specified)
LOG: Appbase = file:///C:/Users/Basic/Downloads/EasyHook/
LOG: Initial PrivatePath = NULL
LOG: Dynamic Base = NULL
LOG: Cache Base = NULL
LOG: AppName = FileMon.exe
Calling assembly : System.Configuration, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a.
===
LOG: This bind starts in default load context.
LOG: No application configuration file found.
LOG: Using host configuration file: 
LOG: Using machine configuration file from C:\Windows\Microsoft.NET\Framework64\v4.0.30319\config\machine.config.
LOG: Binding succeeds. Returns assembly from C:\Windows\Microsoft.Net\assembly\GAC_MSIL\System.Core\v4.0_4.0.0.0__b77a5c561934e089\System.Core.dll.
LOG: Assembly is loaded in default load context.

</pre></html>"

And the output from FileMon:
There was an error while connecting to target:
System.ApplicationException: STATUS_INTERNAL_ERROR: Unknown error in injected C++ completion routine. (Code: 15)
   at EasyHook.RemoteHooking.CreateAndInject(String InEXEPath, String InCommandLine, Int32 InProcessCreationFlags, InjectionOptions InOptions, String InLibraryPath_x86, String InLibraryPath_x64, Int32& OutProcessId, Object[] InPassThruArgs)
   at FileMon.Program.Main(String[] args)
<Press any key to exit>

**Comment by rlc5m8** [comment]
> can you also confirm that you are targetting "AnyCPU" and that you have unticked "Prefer 32-bit".

All Configurations is set to AnyCpu, Prefer-32 Bit is greyed out and unchecked (Seems to be a .NET 4.5 feature). In order to confirm I printed out Environment.Is64BitProcess which returned as true. CorFlags confirmed that the executable is in fact built for x64.

"if you are using the solution from the repository, perhaps try building at the solution level - perhaps it is an assembly version issue?"

I've cloned the repo, as of now I'm up-to-date with the latest commit:
C:\Visual Studio 2015\Projects\EasyHook>git reset --hard origin/master
HEAD is now at fa7d211 Now uses default AppDomain as fallback option (fixes #66)

"1>------ Build started: Project: EasyHook, Configuration: netfx4-Release Any CPU ------
2>------ Build started: Project: ManagedTarget, Configuration: netfx4-Release Any CPU ------
2>  ManagedTarget -> C:\Visual Studio 2015\Projects\EasyHook\Build\netfx4-Release\x64\ManagedTarget.exe
1>  EasyHook -> C:\Visual Studio 2015\Projects\EasyHook\Build\netfx4-Release\x64\EasyHook.dll
3>------ Build started: Project: EasyHookSvc, Configuration: netfx4-Release Any CPU ------
4>------ Build started: Project: ProcessMonitor, Configuration: netfx4-Release Any CPU ------
5>------ Build started: Project: FileMon, Configuration: netfx4-Release Any CPU ------
6>------ Build started: Project: ComplexParameterInject, Configuration: netfx4-Release Any CPU ------
7>------ Build started: Project: ManagedTest, Configuration: netfx4-Release Any CPU ------
8>------ Skipped Build: Project: EasyLoad, Configuration: netfx4-Release x86 ------
8>Project not selected to build for this solution configuration 
9>------ Build started: Project: TestFuncHooks, Configuration: netfx4-Release Any CPU ------
10>------ Build started: Project: EasyHook.Tests, Configuration: netfx4-Release Any CPU ------
11>------ Skipped Build: Project: MultipleHooks (Test\MultipleHooks\MultipleHooks), Configuration: netfx4-Release x86 ------
11>Project not selected to build for this solution configuration 
3>  EasyHookSvc -> C:\Visual Studio 2015\Projects\EasyHook\Build\netfx4-Release\x64\EasyHookSvc.exe
12>------ Skipped Build: Project: EasyHookDll, Configuration: netfx4-Release Win32 ------
12>Project not selected to build for this solution configuration 
13>------ Skipped Build: Project: UnmanagedHook, Configuration: netfx4-Release Win32 ------
13>Project not selected to build for this solution configuration 
14>------ Skipped Build: Project: SimpleHook1, Configuration: netfx4-Release x86 ------
14>Project not selected to build for this solution configuration 
15>------ Skipped Build: Project: SimpleHook2, Configuration: netfx4-Release x86 ------
15>Project not selected to build for this solution configuration 
6>  ComplexParameterInject -> C:\Visual Studio 2015\Projects\EasyHook\Build\netfx4-Release\x64\ComplexParameterInject.dll
10>  EasyHook.Tests -> C:\Visual Studio 2015\Projects\EasyHook\Build\netfx4-Release\x64\EasyHook.Tests.dll
5>  FileMon -> C:\Visual Studio 2015\Projects\EasyHook\Build\netfx4-Release\x64\FileMon.exe
16>------ Build started: Project: ComplexParameterTest, Configuration: netfx4-Release Any CPU ------
7>  ManagedTest -> C:\Visual Studio 2015\Projects\EasyHook\Build\netfx4-Release\x64\ManagedTest.exe
4>  ProcessMonitor -> C:\Visual Studio 2015\Projects\EasyHook\Build\netfx4-Release\x64\ProcessMonitor.exe
17>------ Build started: Project: FileMonInject, Configuration: netfx4-Release Any CPU ------
18>------ Build started: Project: ProcMonInject, Configuration: netfx4-Release Any CPU ------
9>  TestFuncHooks -> C:\Visual Studio 2015\Projects\EasyHook\Build\netfx4-Release\x64\TestFuncHooks.exe
16>  ComplexParameterTest -> C:\Visual Studio 2015\Projects\EasyHook\Build\netfx4-Release\x64\ComplexParameterTest.exe
17>  FileMonInject -> C:\Visual Studio 2015\Projects\EasyHook\Build\netfx4-Release\x64\FileMonInject.dll
18>  ProcMonInject -> C:\Visual Studio 2015\Projects\EasyHook\Build\netfx4-Release\x64\ProcMonInject.dll
========== Build: 12 succeeded, 0 failed, 0 up-to-date, 6 skipped =========="

The binaries all have version "2.7.0.0" instead of "2.7.5870".

Two projects needed to be updated to the new SDK

Issue is resolved after compiling it myself and injection in 32/64-bit process works.
